solo-io/gloo
# Istio 1.18 support

# Description

This PR includes a number of steps required to validate Istio 1.18 support

Fix the istio-proxy sidecar definition in our Helm template, add a new Istio SDS suite to cover the `istioSDS` case, verifying mTLS occurs as expected without auto-injection, 

## Helm changes
- Add required volumes and volume mounts to istio-proxy container in gateway-proxy deployment
  - Augment existing test to ensure these volumes are present 
- Bump default Istio image version value to `1.18.2`

## e2e test changes
- Add mTLS tests to the istio suite with four test cases
  - The suite now installs Gloo with Istio SDS enabled, and autoinjection disabled
  - The mTLS tests apply a deployment of httpbin in the Istio mesh, and makes requests to the `/headers` endpoint
  - Each permutation of peer authentication (strict or permissive) and mTLS on the httpbin upstream (enabled or disabled) is tested

## CI changes
- Bump default Istio version in setup-kind.sh to `1.18.2`
  - Install Istio on `istio-sds` tests 

## Docs changes
- Rewrite Istio guide to set up Istio mTLS using the usage pattern tested in this PR
  - Update versions of k9s, Istio, Gloo to be up-to-date 
- Update Istio support matrix

# Context

Customers have inquired about Istio 1.18 support and would like to see it featured in Gloo 1.16 and 1.15
As such this PR will be back-ported to 1.15

Our existing Istio test doesn't cover the user's case, whereby Istio SDS is enabled, mTLS is required, and Gloo components do not have Istio injection enabled, so in order to claim Istio 1.18 support, we add tests around that case

In addition to writing the e2e, I have manually verified expected behavior following steps analogous to the test (see _Testing steps_)
I have also confirmed that the behavior is the same when the namespace isn't labeled for Istio injection and the respective deployments are patched to enable Istio injection

## Interesting decisions
 
The structure of the test is derived from the [exiting kube2e Istio tests](https://github.com/solo-io/gloo/tree/a8f4a39b4d474b97524d7cbe5e17feb545788754/test/kube2e/istio), with some helper functions left more or less unchanged
Some cleanup/refactoring was done, but was not a high priority

There is an existing [glooctl test](https://github.com/solo-io/gloo/blob/c99cb6ec0109480002529adb75d893e95a7bae2b/test/kube2e/glooctl/istio_test.go#L126) that appears to do something similar, however based on my investigation, it passes despite the gateway-proxy pod not coming up healthily with Istio injected and therefore I have low confidence that it's a strong test
Addressing this is considered out of scope for the PR, as glooctl implements its own logic for Istio injection separately from the Helm chart which we document and prefer

The existing Istio suite did not rely on the Istio injection usage pattern, so it was adapted to match the recommended bespoke Istio sidecar pattern 

The docs no longer reference the Istio injection usage pattern

## Testing steps

### Docs
For the docs guide, follow the guide
I set up a port forward to the `gateway-proxy` rather than using `glooctl proxy url`

### kube2e tests
The the e2e tests, from project root, set up the cluster with:
```
CLUSTER_NODE_VERSION=v1.27.3@sha256:3966ac761ae0136263ffdb6cfd4db23ef8a83cba8a463690e98317add2c9ba72 KUBE2E_TESTS=istio ./ci/kind/setup-kind.sh
```
Run the tests with:
```
go test ./test/kube2e/istio
```

### Manual
To manually verify Istio works correctly with mTLS the e2e test behavior can be replicated by taking the following steps:
Steps:
- Start with a fresh cluster
- Install Istio 1.18 (patch version shouldn't matter)
- Create the gloo-system namespace
- Apply the httpbin app using `kubectl apply -n gloo-system -f https://raw.githubusercontent.com/solo-io/workshops/master/gloo-edge/data/httpbin.yaml`
- Install gloo via helm using [these values](https://github.com/solo-io/gloo/blob/4a47bed3a080673786d4ede9cc993fe19c45118c/test/kube2e/istio-sds/artifacts/helm.yaml)
- Wait for Gloo to come up healthily (should take <90s, but allow a few minutes just in case)
- Use kubectl or k9s to confirm that both the gateway-proxy and httpbin pods have the istio-proxy containers
- Confirm the upstream `gloo-system-httpbin-8000` exists
- Apply the following VS:
```
apiVersion: gateway.solo.io/v1
kind: VirtualService
metadata:
  name: testrunner
  namespace: gloo-system
spec:
  virtualHost:
    domains:
    - '*'
    routes:
    - matchers:
      - prefix: /
      routeAction:
        single:
          upstream:
            name: gloo-system-httpbin-8000
            namespace: gloo-system
```
- Apply the following PeerAuthentication:
```
apiVersion: "security.istio.io/v1beta1"
kind: "PeerAuthentication"
metadata:
  name: "test"
  namespace: "istio-system"
spec:
  mtls:
    mode: STRICT
```
- Use `glooctl` do enable Istio mTLS on the upstream:
```
glooctl istio disable-mtls --upstream httpbin-ns-httpbin-8000
```
- Set up a port forward or otherwise expose port 8080 on the gateway-proxy Service (I forward to 32000)
- Do a curl to the /headers endpoint as follows: `curl -H "Host: testrunner" "localhost:32000/headers"`
- Observe the response with `X-Forwarded-Client-Cert` present

## Notes for reviewers


# Checklist:


BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/8621
